aptos-move/framework/src/natives/signature.rs

```diff
@@ -117,85 +222,135 @@ fn native_bls12381_aggregate_pop_verified_pubkeys(
+ *
+ * where +? indicates that the expression stops evaluating there if the previous gas-charging step
+ * failed
  *
+ * NOTE(ViableSig): We refer to the signatures that deserialize correctly as "viable" above, since
+ * successful deserialization is not a sufficient criteria for the "viability" of a signature: e.g.,
+ * the signature might not verify under the desired (message, public key) pair, or the signature
+ * could lie in a small-order subgroup.
  **************************************************************************************************/
 #[derive(Debug, Clone)]
-pub struct Bls12381VerifyProofOfPosessionGasParameters {
+pub struct Bls12381AggregateSignaturesGasParameters {
     pub base_cost: u64,
+    pub per_sig_deserialize_cost: u64,
+    pub per_sig_aggregate_cost: u64,
 }
 
-fn native_bls12381_verify_proof_of_possession(
-    gas_params: &Bls12381VerifyProofOfPosessionGasParameters,
+pub fn native_bls12381_aggregate_signatures(
+    gas_params: &Bls12381AggregateSignaturesGasParameters,
     _context: &mut NativeContext,
     _ty_args: Vec<Type>,
     mut arguments: VecDeque<Value>,
 ) -> PartialVMResult<NativeResult> {
     debug_assert!(_ty_args.is_empty());
-    debug_assert!(arguments.len() == 2);
+    debug_assert!(arguments.len() == 1);
 
-    let pop_bytes = pop_arg!(arguments, Vec<u8>);
-    let key_bytes = pop_arg!(arguments, Vec<u8>);
+    // Parses a Vec<Vec<u8>> of all serialized signatures
+    let sigs_serialized = pop_vec_arg!(arguments, Vec<u8>);
+    let mut sigs = vec![];
 
-    // NOTE(Gas): O(1) deserialization cost
-    let pop = match bls12381::ProofOfPossession::try_from(&pop_bytes[..]) {
-        Ok(pop) => pop,
-        Err(_) => {
-            return Ok(NativeResult::ok(
-                gas_params.base_cost,
-                smallvec![Value::bool(false)],
-            ))
-        }
-    };
+    let mut cost = gas_params.base_cost;
 
-    // NOTE(Gas): O(1) deserialization cost
-    let public_key = match bls12381::PublicKey::try_from(&key_bytes[..]) {
-        Ok(key) => key,
-        Err(_) => {
-            return Ok(NativeResult::ok(
-                gas_params.base_cost,
-                smallvec![Value::bool(false)],
-            ))
-        }
-    };
+    for sig_bytes in sigs_serialized {
+        // NOTE(Gas): O(1) deserialization cost per signature
+        cost += gas_params.per_sig_deserialize_cost;
 
-    // NOTE(Gas): O(1) cost: 2 bilinear pairings and a hash-to-curve
-    let valid = pop.verify(&public_key).is_ok();
+        let sig = match bls12381::Signature::try_from(&sig_bytes[..]) {
+            Ok(sig) => sig,
+            // If signature does not deserialize correctly, return None.
+            Err(_) => return Ok(NativeResult::ok(cost, smallvec![none_option()])),
+        };
+
+        sigs.push(sig);
+    }
+
+    // If zero signatures were given as input, return None.
+    if sigs.is_empty() {
+        return Ok(NativeResult::ok(cost, smallvec![none_option()]));
+    }
```

Comment:
couldn't we move this up to line 268ish on the input?

Reply:
Sure. Fixed!